fendermint/app/src/main.rs

```diff
@@ -8,61 +8,64 @@ use tracing_appender::{
-use tracing_subscriber::{
-    fmt::{self, writer::MakeWriterExt},
-    layer::SubscriberExt,
-};
+use tracing_subscriber::{fmt, layer::SubscriberExt, Layer};
 
 mod cmd;
 
 fn init_tracing(opts: &options::Options) -> Option<WorkerGuard> {
     let mut guard = None;
 
-    let Some(log_level) = opts.tracing_level() else {
+    let console_filter = opts.log_console_filter().expect("invalid filter");
+    let file_filter = opts.log_file_filter().expect("invalid filter");
+
+    if console_filter.is_none() && file_filter.is_none() {
         return guard;
-    };
+    }
 
     let registry = tracing_subscriber::registry();
 
     // add a file layer if log_dir is set
-    let registry = registry.with(if let Some(log_dir) = &opts.log_dir {
-        let filename = match &opts.log_file_prefix {
-            Some(prefix) => format!("{}-{}", prefix, "fendermint"),
-            None => "fendermint".to_string(),
-        };
-
-        let appender = RollingFileAppender::builder()
-            .filename_prefix(filename)
-            .filename_suffix("log")
-            .rotation(Rotation::DAILY)
-            .max_log_files(5)
-            .build(log_dir)
-            .expect("failed to initialize rolling file appender");
-
-        let (non_blocking, g) = tracing_appender::non_blocking(appender);
-        guard = Some(g);
-
-        Some(
-            fmt::Layer::new()
+    let registry = registry.with(match (&opts.log_dir, file_filter) {
+        (Some(log_dir), Some(filter)) => {
+            let filename = match &opts.log_file_prefix {
+                Some(prefix) => format!("{}-{}", prefix, "fendermint"),
+                None => "fendermint".to_string(),
+            };
+
+            let appender = RollingFileAppender::builder()
+                .filename_prefix(filename)
+                .filename_suffix("log")
+                .rotation(Rotation::DAILY)
+                .max_log_files(7)
+                .build(log_dir)
+                .expect("failed to initialize rolling file appender");
+
+            let (non_blocking, g) = tracing_appender::non_blocking(appender);
+            guard = Some(g);
+
+            let file_layer = fmt::Layer::new()
                 .json()
-                .with_writer(non_blocking.with_max_level(log_level))
+                .with_writer(non_blocking)
                 .with_span_events(FmtSpan::CLOSE)
                 .with_target(false)
                 .with_file(true)
-                .with_line_number(true),
-        )
-    } else {
-        None
+                .with_line_number(true)
+                .with_filter(filter);
+
+            Some(file_layer)
+        }
+        _ => None,
     });
 
-    // we also log all traces with level INFO or higher to stdout
-    let registry = registry.with(
+    // we also log all traces to stderr (reserving stdout for any actual output such as from the CLI commands)
+    let registry = registry.with(console_filter.map(|filter| {
```

Comment:
You are right, apparently you can! However I think that would not be what `None` was supposed to mean, which is basically `OFF`. I suspect if we call `.with_filter(None)` then the effect is like not calling the method at all, which means it will log something, whereas I wanted to say that if you configure `"off"` which is turned into `None` then there are no logs at all. 

This is where we could simplify the thing and just return a non-optional `EnvFilter`. I didn't know what was better: not configuring anything, or configuring something with and `off` filter. 

But if I'm honest we'll probably never have a case of not logging anything at all, so we might as well simplify what we can, since this part of the code is just more complicated to maintain for little benefit.